pkg/datagatherer/versionchecker/versionchecker.go

```diff
@@ -0,0 +1,256 @@
+				Bearer:   data["bearer"],
+				Host:     data["host"],
+			}
+
+			c.VersionCheckerClientOptions.Selfhosted[fmt.Sprintf("selfhosted-%d", i+1)] = &opts
+		default:
+			return fmt.Errorf("registry %d/%d was an unknown kind (%s)", i+1, len(aux.Registries), v.Kind)
+		}
+	}
+
+	for k, v := range registryKindCounts {
+		if v > 1 && k != "selfhosted" {
+			return fmt.Errorf("found %d registries of kind %s, only 1 is supported", v, k)
+		}
+	}
+
+	return nil
+}
+
+func loadKeysFromPaths(keys []string, params map[string]string) (map[string]string, error) {
+	requiredKeys := map[string]bool{}
+	for _, v := range keys {
+		requiredKeys[v] = false
+	}
+
+	loadedData := map[string]string{}
+	for _, k := range keys {
+		path := params[k]
+		if path == "" {
+			// don't try to load unset secrets. version-checker will fail if
+			// config is missing
+			continue
+		}
+
+		file, err := os.Open(path)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load file for %s at %s: %s", k, path, err)
+		}
+		defer file.Close()
+
+		b, err := ioutil.ReadAll(file)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read file for %s at %s: %s", k, path, err)
+		}
+		loadedData[k] = strings.TrimSpace(string(b))
+	}
+
+	return loadedData, nil
+}
+
+// NewDataGatherer creates a new VersionChecker DataGatherer
+func (c *Config) NewDataGatherer(ctx context.Context) (datagatherer.DataGatherer, error) {
+	// create the k8s DataGatherer to use when collecting pods
+	dynamicDg, err := c.Dynamic.NewDataGatherer(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// configure version checker parameters
+	vclog := logrus.New()
+	vclog.SetOutput(os.Stdout)
+	log := logrus.NewEntry(vclog)
+	imageClient, err := vcclient.New(ctx, log, c.VersionCheckerClientOptions)
+	if err != nil {
+		return nil, fmt.Errorf("failed to setup version checker image registry clients: %s", err)
+	}
+	timeout := 30 * time.Minute // this is the default used in version checker
+	search := vcsearch.New(
+		log,
+		timeout,
+		vcversion.New(log, imageClient, timeout),
+	)
+
+	// dg wraps version checker and dynamic client to request pods
+	return &DataGatherer{
+		ctx:                   ctx,
+		config:                c,
+		dynamicDg:             dynamicDg,
+		versionChecker:        vcchecker.New(search),
+		versionCheckerLog:     log,
+		versionCheckerOptions: c.VersionCheckerCheckerOptions,
+	}, nil
+}
+
+// DataGatherer is a VersionChecker DataGatherer
+type DataGatherer struct {
+	ctx                   context.Context
+	config                *Config
+	dynamicDg             datagatherer.DataGatherer
+	versionChecker        *vcchecker.Checker
+	versionCheckerLog     *logrus.Entry
+	versionCheckerOptions vcapi.Options
+}
+
+// PodResult wraps a pod and a version checker result for an image found in one
+// of the containers for that pod. Exported so the backend can destructure
+// json.
+type PodResult struct {
+	Pod    v1.Pod            `json:"pod"`
+	Result *vcchecker.Result `json:"result"`
+}
```

Comment:
Are we duplicating the pod for every container we use? Could we improve this by wrapping the results with every paired containes?

```yaml
result:
  pod: {}
  results:
    containerName: &Result{}
```

Reply:
I've added something similar in 22023da7da642e2e50a1668fce87517e56938a7b but it also tracks if the container is an init container. (I don't hate maps, honest :joy: )